config/application.rb

```diff
@@ -59,6 +59,22 @@ class Application < Rails::Application
     # Therefore we disable "per_form_csrf_tokens" for the time being.
     config.action_controller.per_form_csrf_tokens = false
 
+    # Enable CORS for API v0
+    # (logging is only activated when debug is enabled)
+    config.middleware.insert_before 0, Rack::Cors, debug: Rails.env.development?, logger: (-> { Rails.logger }) do
+      allow do
+        origins do |source, _env|
+          source # echo back the client's `Origin` header instead of using `*`
+        end
+
+        # allowed public APIs
+        %w[articles comments listings podcast_episodes tags users videos].each do |resource_name|
+          # allow read operations, disallow custom headers (eg. api-key) and disable preflight caching
```

Comment:
I'll enable preflight caching in a second moment, as browsers can cache between 0 to 24hours, when we deploy this I'll make sure to test all endpoints, once I confirm they all work correctly, I'll enable a suitable caching time

Reply:
>  disallow custom headers (eg. api-key)

It feels like maybe we should allow `api-key` so users can switch between GET and PUT/POST/DELETE requests without worrying about stripping that header first. What do you think? I feel like those with scripts already written are going to be frustrated when they blow up.

Reply:
Thanks for the comment. Server side scripts are not an issue, what CORS blocks is in browser JS. As I said in the description:

> I purposefully left out unauthenticated endpoints because I want to dig deeper in the security implications (in addition to the fact that it'd encourage people to build frontends that might leak DEV API keys).

This is exactly the discussion I wanted to spark. I'm not super pro on allowing CORS for frontend calls on protected routes because it encourages people to hardcode api Keys in their frontends. The CORS configuration we have right now it's not correct, hence why issues like #5877 exists.

It probably should be a team decision, but anyhow I'd rather enable it in steps (it's literally a one line change) if we decide to do that.

Reply:
Gotcha! Thanks for the explanation, yeah I am not a CORS expert either.